cmd/crank/beta/diff/client/kubernetes/schema_client.go

```diff
@@ -0,0 +1,146 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	un "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/cmd/crank/beta/diff/client/core"
+)
+
+// SchemaClient handles operations related to Kubernetes schemas and CRDs.
+type SchemaClient interface {
+	// GetCRD gets the CustomResourceDefinition for a given GVK
+	GetCRD(ctx context.Context, gvk schema.GroupVersionKind) (*un.Unstructured, error)
+
+	// IsCRDRequired checks if a GVK requires a CRD
+	IsCRDRequired(ctx context.Context, gvk schema.GroupVersionKind) bool
+
+	// ValidateResource validates a resource against its schema
+	ValidateResource(ctx context.Context, resource *un.Unstructured) error
+}
+
+// DefaultSchemaClient implements SchemaClient.
+type DefaultSchemaClient struct {
+	dynamicClient dynamic.Interface
+	typeConverter TypeConverter
+	logger        logging.Logger
+
+	// Resource type caching
+	resourceTypeMap map[schema.GroupVersionKind]bool
+	resourceMapMu   sync.RWMutex
+}
+
+// NewSchemaClient creates a new DefaultSchemaClient.
+func NewSchemaClient(clients *core.Clients, typeConverter TypeConverter, logger logging.Logger) SchemaClient {
+	return &DefaultSchemaClient{
+		dynamicClient:   clients.Dynamic,
+		typeConverter:   typeConverter,
+		logger:          logger,
+		resourceTypeMap: make(map[schema.GroupVersionKind]bool),
+	}
+}
+
+// GetCRD gets the CustomResourceDefinition for a given GVK.
+func (c *DefaultSchemaClient) GetCRD(ctx context.Context, gvk schema.GroupVersionKind) (*un.Unstructured, error) {
+	// Get the pluralized resource name
+	resourceName, err := c.typeConverter.GetResourceNameForGVK(ctx, gvk)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot determine CRD name for %s", gvk.String())
+	}
+
+	c.logger.Debug("Looking up CRD", "gvk", gvk.String(), "crdName", resourceName)
+
+	// Construct the CRD name using the resource name and group
+	crdName := fmt.Sprintf("%s.%s", resourceName, gvk.Group)
+
+	// Define the CRD GVR directly to avoid recursion
+	crdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	// Fetch the CRD
+	crd, err := c.dynamicClient.Resource(crdGVR).Get(ctx, crdName, metav1.GetOptions{})
+	if err != nil {
+		c.logger.Debug("Failed to get CRD", "gvk", gvk.String(), "crdName", crdName, "error", err)
+		return nil, errors.Wrapf(err, "cannot get CRD %s for %s", crdName, gvk.String())
+	}
+
+	c.logger.Debug("Successfully retrieved CRD", "gvk", gvk.String(), "crdName", resourceName)
+	return crd, nil
+}
+
+// IsCRDRequired checks if a GVK requires a CRD.
```

Comment:
Do you really need CRDs specifically, or could you use the Kubernetes discovery client?

I ask because the discovery client should be able to get you things like schemas (and whether a type even exists) for both built-in and custom types, so you wouldn't need to worry about whether a type has a CRD or not.